python/mrc/_pymrc/include/pymrc/edge_adapter.hpp

```diff
@@ -59,141 +61,94 @@ namespace mrc::pymrc {
+    using source_adapter_fn_t = std::function<std::shared_ptr<node::IEdgeWritableBase>(
+        mrc::node::SourcePropertiesBase&, mrc::node::SinkPropertiesBase&, std::shared_ptr<node::IEdgeWritableBase>)>;
+    using sink_adapter_fn_t   = std::function<std::shared_ptr<node::IEdgeWritableBase>(
+        std::type_index, mrc::node::SinkPropertiesBase&, std::shared_ptr<node::IEdgeWritableBase> ingress_handle)>;
+
+    using ingress_adapter_fn_t = node::EdgeAdapterRegistry::ingress_adapter_fn_t;
 
     template <typename DataTypeT>
     static void register_data_adapters()
     {
-        if (!mrc::node::EdgeAdapterRegistry::has_source_adapter(typeid(DataTypeT)))
-        {
-            std::type_index source_type = typeid(DataTypeT);
-            VLOG(2) << "Registering PyMRC source adapter for: " << type_name<DataTypeT>() << " "
-                    << source_type.hash_code();
-            node::EdgeAdapterRegistry::register_source_adapter(typeid(DataTypeT),
-                                                               EdgeAdapterUtil::build_source_adapter<DataTypeT>());
-        }
-
-        if (!mrc::node::EdgeAdapterRegistry::has_sink_adapter(typeid(DataTypeT)))
-        {
-            std::type_index sink_type = typeid(DataTypeT);
-            VLOG(2) << "Registering PyMRC sink adapter for: " << type_name<DataTypeT>() << " " << sink_type.hash_code();
-            node::EdgeAdapterRegistry::register_sink_adapter(typeid(DataTypeT),
-                                                             EdgeAdapterUtil::build_sink_adapter<DataTypeT>());
-        }
+        node::EdgeAdapterRegistry::register_ingress_adapter(EdgeAdapterUtil::build_sink_ingress_adapter<DataTypeT>());
+        node::EdgeAdapterRegistry::register_ingress_adapter(EdgeAdapterUtil::build_source_ingress_adapter<DataTypeT>());
     }
 
     template <typename InputT>
-    static sink_adapter_fn_t build_sink_adapter()
+    static ingress_adapter_fn_t build_sink_ingress_adapter()
     {
-        return [](std::type_index source_type,
-                  mrc::node::SinkPropertiesBase& sink,
-                  std::shared_ptr<channel::IngressHandle> ingress_handle) {
-            if (source_type == typeid(PyHolder))
+        return [](const node::EdgeTypePair& target_type, std::shared_ptr<node::IEdgeWritableBase> ingress_handle) {
+            // First try to convert the ingress to our type
+            auto typed_ingress = std::dynamic_pointer_cast<node::IEdgeWritable<InputT>>(ingress_handle);
+
+            if (!typed_ingress)
+            {
+                // Cant do anything about this ingress
+                return std::shared_ptr<node::IngressHandleObj>(nullptr);
+            }
+
+            auto ingress_type = ingress_handle->get_type();
+
+            // Check to see if we have a conversion in pybind11
+            if (pybind11::detail::get_type_info(ingress_type.unwrapped_type(), false))
```

Comment:
This actually ends up being problematic, because the internal data structure that Pybind uses is specific to the PYBIND11_INTERNALS_ID value, meaning if you call this in Morpheus for a type that was registered in MRC, it may not find it and you'll always fall back to the Pyobject case (https://github.com/nv-morpheus/MRC/issues/176). We're probably going to want an outside registry lookup to map to the correct pybind11 object.

Check pybind/detail/internal.h:270
```c++
#define PYBIND11_INTERNALS_ID "__pybind11_internals_v" \
    PYBIND11_TOSTRING(PYBIND11_INTERNALS_VERSION) PYBIND11_INTERNALS_KIND PYBIND11_COMPILER_TYPE PYBIND11_STDLIB PYBIND11_BUILD_ABI PYBIND11_BUILD_TYPE "__"

#define PYBIND11_MODULE_LOCAL_ID "__pybind11_module_local_v" \
    PYBIND11_TOSTRING(PYBIND11_INTERNALS_VERSION) PYBIND11_INTERNALS_KIND PYBIND11_COMPILER_TYPE PYBIND11_STDLIB PYBIND11_BUILD_ABI PYBIND11_BUILD_TYPE "__"

/// Each module locally stores a pointer to the `internals` data. The data
/// itself is shared among modules with the same `PYBIND11_INTERNALS_ID`.
inline internals **&get_internals_pp() {
    static internals **internals_pp = nullptr;
    return internals_pp;
}
```

Reply:
I agree with you that we will need to address this issue but this code existed before and the functionality hasnt changed. Only the file has been changed and the location of where this lookup happens.

Seems like we will likely need to do something similar to the Automatic type registration that we did for ports.